### PR TITLE
[AI-3] Store the model catalogue and capability verdicts

### PR DESCRIPTION
**Pull request by tangopium**
# Ticket

[AI-3](https://community.openproject.org/work_packages/66020)

# What are you trying to accomplish?

PR 4 of 11 in the AI-3 stack. Persists what the server offers and what is known about it: `LlmModel` rows synced from the catalogue after every credential change and on demand via the Refresh models action, and `LlmCapabilityVerdict` rows in three states (supported, unsupported, unknown), because the OpenAI model list carries no capability information and "we cannot tell" is the honest answer. Verdicts are seeded from the ruby_llm registry where the model id is catalogued, administrator assertions survive re-detection, and everything is discarded when the credentials point at a different deployment. Models the server stops offering are deactivated rather than deleted, so anything referencing one still has something to name. No UI beyond the refresh action; the list follows in PR 5.

# Merge checklist

- [x] Added/updated tests
- [x] Behind the `llm_connection` feature flag

Stacked on #24881.

### CODE REVIEW

File: app/services/llm_connections/sync_models_service.rb [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+  #
+  # Kept separate from the contract probe so the same path serves the "Refresh
+  # models" button, the update service and the environment seeder.
+  class SyncModelsService
+    def initialize(connection)
+      @connection = connection
+    end
+
+    def call
+      store(adapter.models)
```
[the remark this answers is not in the document]
Comment (tangopium) [reply app/services/llm_connections/sync_models_service.rb:47]: Fixed: the fingerprint comparison now runs before the fetch and deactivates the old models as well as deleting the verdicts, so a deployment change invalidates the old state even when the new server refuses the model list. Covered by a new service spec.

File: app/services/llm_connections/sync_models_service.rb [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+    end
+
+    def upsert(cards)
+      now = Time.current
+
+      cards.each do |card|
+        model = connection.models.find_or_initialize_by(external_id: card.fetch(:id))
+        model.update!(display_name: card[:display_name],
+                      raw_metadata: card.fetch(:raw, {}),
+                      last_seen_at: now,
```
[the remark this answers is not in the document]
Comment (tangopium) [reply app/services/llm_connections/sync_models_service.rb:96]: Fixed: the upsert now merges admin_context_window back into the server card instead of replacing the hash wholesale, with a spec.

File: app/services/llm_connections/sync_models_service.rb [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+    end
+
+    # Same deployment, but a model is gone. Its verdict is meaningless now,
+    # except an administrator's assertion: an operator restarting a server must
+    # not silently lose one.
+    def discard_verdicts_for_vanished_models
+      known = connection.available_model_ids
+      return if known.empty?
+
+      connection.capability_verdicts.where.not(model_id: known).where.not(source: "admin").delete_all
```
[the remark this answers is not in the document]
Comment (tangopium) [reply app/services/llm_connections/sync_models_service.rb]: Fixed: an empty catalogue now deletes every non-admin verdict. This also surfaced the same where.not-empty-array pitfall in withdraw_models_absent_from, which is fixed too, with a spec for the empty-catalogue case.

File: app/models/llm_connection.rb [hunk 4]
```diff
@@ -38,9 +38,8 @@ class LlmConnection < ApplicationRecord
   include Redmine::Ciphering
 
   SINGLETON_NAME = "default"
-
-  has_many :health_reports, as: :subject, dependent: :delete_all
-
+  has_many :models, class_name: "LlmModel", dependent: :delete_all
+  has_many :capability_verdicts, class_name: "LlmCapabilityVerdict", dependent: :delete_all
```
[the remark this answers is not in the document]
Comment (tangopium) [reply app/models/llm_connection.rb:43]: Fixed: the health_reports association stays on the model from part 1 onward.

File: app/workers/llm/sync_models_job.rb [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+#++
+
+module Llm
+  # Refreshes the cached model catalogue out of band.
+  #
+  # Used by the environment seeder, which must not block on -- or fail because of
+  # -- an LLM server that has not finished starting.
+  class SyncModelsJob < ApplicationJob
+    def perform
+      connection = LlmConnection.first
```
Comment (thykel) [inline app/workers/llm/sync_models_job.rb:38]: I don't see a reason to implement that single-connection constraint here _in addition_ to the model. We can just iterate the whole table to reduce friction later.

File: app/services/llm_connections/sync_models_service.rb [hunk 6]
```diff
@@ -0,0 +1,152 @@
...
+    def fingerprint
+      @fingerprint ||= Digest::SHA256.hexdigest("#{connection.api_format}\0#{connection.base_url}\0#{connection.api_key}")
+    end
+
+    def upsert(cards)
+      now = Time.current
+
+      cards.each do |card|
+        model = connection.models.find_or_initialize_by(external_id: card.fetch(:id))
+        model.update!(display_name: card[:display_name],
```
Comment (thykel) [inline app/services/llm_connections/sync_models_service.rb:94]: Shouldn't we want to also retain the original name if we got nothing from the server? Something like:

```suggestion
        model.update!(display_name: card[:display_name].presence || model.display_name),
```

File: app/services/llm/capabilities.rb [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+      nil
+    end
+
+    def states_from(info)
+      embedding = info.type.to_s == "embedding"
+      published = Array(info.capabilities).map(&:to_sym)
+      relevant = embedding ? EMBEDDING : CHAT
+
+      states = relevant.index_with { |capability| published.include?(capability) ? :supported : :unsupported }
+      states.merge(embeddings: embedding ? :supported : :unsupported)
```
Comment (thykel) [inline app/services/llm/capabilities.rb:72]: Aren't we missing the edge case of `unknown` here?

**Review by thykel — COMMENTED** [review]
👋 The concept looks sound to me, but a couple of things leave me scratching my head. Additionally, this needs way more **test coverage** for the new models, sync service, and enrichment mechanism.

**Comment by tangopium** [comment]
The branch was rebuilt, so the commits your comments point at no longer exist and those threads now render as outdated. Nothing was dropped. Here is where each point landed.

- The refresh is an initial fill only, so a re-seed no longer overwrites a curated list.
- Administrator edits survive a refresh.
- A model's type is derived from its embeddings verdict.
- The connection defaults are now real references into `llm_models` rather than identifier strings, added here because the table has to exist first. The foreign keys clear a default when its model is deleted.

Still open on this PR: the broader coverage you asked for on the new models, the sync service and the enrichment mechanism is not in yet. It exists on a separate branch and has not been moved across.

CI has not started on the new commits: only the CLA check ran, and the workflow runs seem to need maintainer approval on this repository. The checks are waiting on that.

**Review by akabiru — COMMENTED** [review]
@tangopium before I have a look please address the previous review from Tom

File: app/services/llm_connections/update_service.rb [hunk 8]
```diff
@@ -32,10 +32,25 @@ module LlmConnections
...
+    # the save. A sync failure is therefore logged, not surfaced.
+    def after_perform(service_call)
+      super.tap do
+        next unless service_call.success?
 
-      call
+        Setting.llm_features_enabled = model.llm_features_enabled
+        next unless initial_fill?(service_call.result)
+
+        SyncModelsService.new(service_call.result).call
```
Comment (akabiru) [inline app/services/llm_connections/update_service.rb:45]: 🍎 **The catalogue sync runs a live HTTP call inside the service transaction and advisory lock.**

`after_perform` is called from `BaseContracted#perform` inside `service_context`, and `Shared::ServiceContext#in_mutex_context` wraps that in `OpenProject::Mutex.with_advisory_lock_transaction`, which is `transaction { with_advisory_lock { yield } }`. So this call, up to 20 seconds of network time with the client's probe timeout, holds an open DB transaction and the advisory lock on the connection row. `Llm::SyncModelsJob`'s own comment exists to keep slow servers out of exactly this position, and the initial connect is where a self-hosted server is most likely to be slow.

Move the call out of the service: in the controller, between `result.on_success` and `redirect_after_save`, still synchronous because the flash reads `models.none?` right after. Not an `after_commit` on the model, which only fires on the outermost transaction and breaks under the transactional request specs.

File: app/services/llm_connections/sync_models_service.rb [hunk 9]
```diff
@@ -0,0 +1,154 @@
...
+    end
+
+    # The server names the model, but only when it says so: the administrator's
+    # display name and context-window override are theirs, and a routine refresh
+    # must not silently discard them.
+    def merged_metadata(model, card)
+      raw = card.fetch(:raw, {})
+      admin_window = model.raw_metadata["admin_context_window"]
+
+      admin_window ? raw.merge("admin_context_window" => admin_window) : raw
```
Comment (akabiru) [inline app/services/llm_connections/sync_models_service.rb:140]: 🍊 **A remote card can forge an administrator's context window.**

`raw` is the whole card as the server sent it. The stored `admin_context_window` is re-injected only when the model already has one, and the incoming key is never stripped. A server that includes `"admin_context_window"` in its card gets it written verbatim on first sync, `LlmModel#context_window_source` reports `:admin` because the key is present, and from then on this method treats it as the real override. No spec covers a card carrying that key.

```suggestion
      admin_window ? raw.except("admin_context_window").merge("admin_context_window" => admin_window) : raw.except("admin_context_window")
```

Or nest remote data under its own key so remote and admin-owned keys can never collide; that also helps #24883 and #24884, which read this hash.

File: app/services/llm_connections/sync_models_service.rb [hunk 10]
```diff
@@ -0,0 +1,154 @@
...
+    def call
+      # Before the fetch, so that a deployment change invalidates the old state
+      # even when the new server refuses the model list: keeping the previous
+      # deployment's models and verdicts under new credentials would be wrong.
+      invalidate_a_different_deployment
+
+      store(adapter.models)
+
+      ServiceResult.success(result: connection)
+    rescue Llm::Client::Error => e
```
Comment (akabiru) [inline app/services/llm_connections/sync_models_service.rb:50]: 🍊 **Only `Llm::Client::Error` is rescued; anything the store raises escapes as a 500.**

`external_id` has no length bound, so a card with a multi-kilobyte `id` exceeds the btree index entry limit and `update!` raises `StatementInvalid`. Two overlapping syncs race `find_or_initialize_by` then `update!` into `RecordNotUnique`. Neither is caught here, so on the save path it is a 500 and in the job it aborts every connection after the failing one. Nothing caps the number of cards or the size of `raw` either; a server listing a hundred thousand models is parsed whole and written row by row inside one transaction.

Bound the input and widen the net: `validates :external_id, length: { maximum: 512 }` on the model, drop oversized ids at the adapter, cap the cards at a constant and log the truncation, and rescue `ActiveRecord::ActiveRecordError` here into `ServiceResult.failure`.

File: app/workers/llm/sync_models_job.rb [hunk 11]
```diff
@@ -0,0 +1,43 @@
...
+
+module Llm
+  # Refreshes the cached model catalogue out of band.
+  #
+  # Used by the environment seeder, which must not block on -- or fail because of
+  # -- an LLM server that has not finished starting.
+  class SyncModelsJob < ApplicationJob
+    def perform
+      LlmConnection.find_each do |connection|
+        LlmConnections::SyncModelsService.new(connection).call
```
Comment (akabiru) [inline app/workers/llm/sync_models_job.rb:39]: 🍊 **One failing connection aborts the sync for every connection after it.**

`call` only rescues client errors, so `UnsupportedFormat` from a stale `api_format`, or a `RecordInvalid` from the store, escapes the block and ends the `find_each`. One active row today, but the model documents more than one row as a legitimate state, and #24887 adds a `raise` inside this loop.

```suggestion
      LlmConnection.find_each do |connection|
        LlmConnections::SyncModelsService.new(connection).call
      rescue StandardError => e
        Rails.logger.error { "LLM model sync failed for connection #{connection.id}: #{e.class}" }
      end
```

File: app/workers/llm/sync_models_job.rb [hunk 12]
```diff
@@ -0,0 +1,43 @@
...
+# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
+#
+# See COPYRIGHT and LICENSE files for more details.
+#++
+
+module Llm
+  # Refreshes the cached model catalogue out of band.
+  #
+  # Used by the environment seeder, which must not block on -- or fail because of
+  # -- an LLM server that has not finished starting.
```
Comment (akabiru) [inline app/workers/llm/sync_models_job.rb]: 🍏 The seeder this comment names arrives in #24887; at this head the job has no caller. Describe what the job does, or say it is for a later PR.

File: app/services/llm_connections/enrich_capabilities_service.rb [hunk 13]
```diff
@@ -0,0 +1,85 @@
...
+      end
+
+      llm_model.update!(attributes) if attributes.any?
+    end
+
+    def record(model_id, capability, state)
+      verdict = connection.capability_verdicts.find_or_initialize_by(model_id:, capability: capability.to_s)
+      # Anything an administrator or a probe established beats a published claim:
+      # both looked at this deployment, the registry did not.
+      return if verdict.persisted? && verdict.source.in?(%w[admin probe])
```
Comment (akabiru) [inline app/services/llm_connections/enrich_capabilities_service.rb:80]: 🍊 **The precedence guard omits the `observed` source.** `LlmCapabilityVerdict` defines four sources and this protects two; a registry re-enrichment would overwrite an observed verdict with a published claim, which the comment above says must not happen. Nothing produces `observed` yet, so either add it here or leave the enum value out until something does.

```suggestion
      return if verdict.persisted? && verdict.source.in?(%w[admin probe observed])
```

File: app/models/llm_connection.rb [hunk 14]
```diff
@@ -76,6 +81,28 @@ def configured?
...
   end
 
+  # Every model that can be addressed today: discovered and still offered, plus
+  # anything an administrator entered by hand.
+  #
+  # Deliberately includes models an administrator has deactivated. This is what
+  # Llm::Runtime resolves against, and hiding a model from the pickers must not
+  # break a feature that is already bound to it.
+  def available_model_ids
+    models.active.by_identifier.pluck(:external_id)
```
Comment (akabiru) [inline app/models/llm_connection.rb:91]: 🍊 **The comment says deactivated models are included; the scope excludes them.** `models.active` is `where(active: true)`, and at this head there is no other deactivation state. The factory's `:deactivated` trait sets a `deactivated_at` column that this migration does not create. Both describe the curation state #24885 adds; either move the comment and trait to that PR or land the column here.

File: spec/factories/llm_model_factory.rb [hunk 15]
```diff
@@ -0,0 +1,53 @@
...
+      last_seen_at { nil }
+    end
+
+    trait :withdrawn do
+      active { false }
+    end
+
+    # Still offered by the server; hidden by an administrator.
+    trait :deactivated do
+      deactivated_at { Time.current }
```
Comment (akabiru) [inline spec/factories/llm_model_factory.rb]: 🍏 `deactivated_at` does not exist on `llm_models` until #24885; this trait raises if anything calls it at this head.

File: db/migrate/20260812090000_create_llm_models.rb [hunk 16]
```diff
@@ -0,0 +1,67 @@
...
+    add_index :llm_models, %i[llm_connection_id external_id], unique: true
+
+    # The catalogue is superseded by the table above, and a model the server
+    # stops reporting keeps its row here, so a default can be a reference rather
+    # than an identifier that nothing resolves. Clearing it on delete is what the
+    # administrator is warned about before removing a model.
+    change_table :llm_connections, bulk: true do |t|
+      t.remove :catalogue, type: :jsonb, null: false, default: {}
+      t.remove :default_chat_model_id, type: :string
+      t.remove :default_embedding_model_id, type: :string
```
Comment (akabiru) [inline db/migrate/20260812090000_create_llm_models.rb:60]: 🍊 **The stack creates these three columns in #24879 and drops them here.** Neither migration has shipped, so editing the first one leaves one clean create instead of a create-then-drop pair. `catalogue_fetched_at` also stays behind although the `catalogue` column it was named after goes; `last_synced_at` would say what it means.

File: spec/services/llm_connections/sync_models_service_spec.rb [hunk 17]
```diff
@@ -0,0 +1,129 @@
+# frozen_string_literal: true
```
Comment (akabiru) [inline spec/services/llm_connections/sync_models_service_spec.rb:1]: 🍊 **The everyday branches are untested.** A deployment change and an empty catalogue are covered; a second refresh on the same server that returns one model fewer is not, so `withdraw_models_absent_from` with a non-empty list and the matching verdict cleanup never run under test. One example: two models on the first sync, one on the second, assert the missing one is deactivated and its non-admin verdict dropped while the sibling's survives.

File: app/workers/llm/sync_models_job.rb (hunk 5 above)
Comment (akabiru) [reply app/workers/llm/sync_models_job.rb]: Fixed at head: `perform` is a plain `LlmConnection.find_each` with no single-connection guard (sync_models_job.rb:37-41). Can be resolved. One follow-on in the new comment on line 39: the loop has no per-row rescue, so a failure on one row now skips the rest.

File: app/services/llm_connections/sync_models_service.rb (hunk 6 above)
Comment (akabiru) [reply app/services/llm_connections/sync_models_service.rb]: Fixed at head exactly as suggested: `display_name: card[:display_name].presence || model.display_name` (sync_models_service.rb:94), covered by "keeps an administrator's display name when the server reports none". Can be resolved.

File: app/services/llm/capabilities.rb (hunk 7 above)
Comment (akabiru) [reply app/services/llm/capabilities.rb]: Fixed at head: `states_from` sets `unpublished = published.empty? ? :unknown : :unsupported` (capabilities.rb:66-76), with the spec at capabilities_spec.rb:57-67. Can be resolved.

File: app/services/llm_connections/sync_models_service.rb (hunk 2 above)
[the remark this answers is not in the document]
Comment (akabiru) [reply app/services/llm_connections/sync_models_service.rb]: Fixed for a stored value, but the incoming card is never stripped of the same key, so a server that sends `"admin_context_window"` gets it written as an admin override on first sync. Details and a one-line change in the new comment on `merged_metadata` (line 140).

File: app/workers/llm/sync_models_job.rb (hunk 11 above)
Comment (tangopium) [reply app/workers/llm/sync_models_job.rb]: Done in baa7a02. In #24887 the job now tries every connection before it raises for the retry.

File: app/models/llm_connection.rb (hunk 14 above)
Comment (tangopium) [reply app/models/llm_connection.rb]: Removed both in faa34a4 rather than moving them to #24885, since that state no longer exists anywhere in the stack.

File: app/services/llm_connections/sync_models_service.rb (hunk 10 above)
Comment (tangopium) [reply app/services/llm_connections/sync_models_service.rb]: Oversized ids are now skipped with a warning instead of failing the refresh (dedcd6e); the 2,000-card cap and the ActiveRecord rescue came earlier in baa7a02.

File: db/migrate/20260812090000_create_llm_models.rb (hunk 16 above)
Comment (tangopium) [reply db/migrate/20260812090000_create_llm_models.rb]: Done: #24879 no longer creates catalogue or the string default columns (09adc28), this migration only adds the references (c59608b), and catalogue_fetched_at is now last_synced_at.
